src/test/java/part2/exercise/CollectorsExercise2.java

```diff
@@ -163,95 +163,105 @@ public MapPair(Map<String, Key> keyById, Map<String, List<Value>> valueById) {
         };
     }
 
+    private static Map<Key, List<Value>> collectKeyValueMapWithoutStream(List<Pair> pairs) {
+        final Map<Key, List<Value>> map = new HashMap<>();
+
+        for (Pair pair : pairs) {
+            final Key key = pair.getKey();
+            final Value value = pair.getValue();
+            if (!map.containsKey(key))
+                map.put(key, new ArrayList<>());
+            map.get(key).add(value);
+        }
+
+        return map;
+    }
+
+    private static Map<Key, List<Value>> collectKeyValueMapDoubleIteration(List<Pair> pairs) {
+
+        final Map<String, Key> keyMap =
+                pairs.stream()
+                .collect(
+                        toMap(p -> p.getKey().getId(), Pair::getKey, (a, b) -> a));
+
+        final Map<String, List<Value>> valuesMap = pairs.stream()
+                .collect(
+                        groupingBy(p -> p.getKey().getId(), mapping(Pair::getValue, toList()))
+                );
+
+        return valuesMap.entrySet().stream()
+                .collect(
+                        toMap(e -> keyMap.get(e.getKey()), Map.Entry::getValue));
+
+    }
+
+    private static class MapPairCollector implements Collector<Pair, MapPair, MapPair> {
+
+        @Override
+        public Supplier<MapPair> supplier() {
+            return MapPair::new;
+        }
+
+        @Override
+        public BiConsumer<MapPair, Pair> accumulator() {
+            return (m, p) -> {
+                final Key key = p.getKey();
+                final String sKey = key.toString();
+                final Value value = p.getValue();
+                final Map<String, Key> keyMap = m.getKeyById();
+                final Map<String, List<Value>> valueMap = m.getValueById();
+                keyMap.put(sKey, key);
+                if (!valueMap.containsKey(sKey))
+                    valueMap.put(sKey, new ArrayList<>());
+                valueMap.get(sKey).add(value);

```

Comment:
`Key` and `Value` in `Pair` are not related. You should use `Key::getId` and `Value::getKeyId`